File: Player/Windows/AndoW_Player/Services/SignalRClientService.cs
using System;
using AndoW.Shared;
using System.Collections.Generic;
using System.Configuration;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.AspNet.SignalR.Client;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using TurtleTools;
using SharedWeeklyPlayScheduleInfo = AndoW.Shared.WeeklyPlayScheduleInfo;

namespace HyOnPlayer
{
    internal sealed class SignalRClientService : IDisposable
    {
        private const int DefaultPort = 5000;
        private const string DefaultHubPath = "/Data";
        private const int ReconnectDelayMs = 5000;

        private readonly MainWindow owner;
        private readonly RemoteCommandService commandService;
        private readonly object syncRoot = new object();
        private readonly ServerSettingsClient serverSettingsClient;
        private HubConnection connection;
        private IHubProxy hubProxy;
        private int reconnecting;
        private int stopping;
        private string currentUrl;

        public SignalRClientService(MainWindow owner, RemoteCommandService commandService)
        {
            this.owner = owner;
            this.commandService = commandService;
            serverSettingsClient = new ServerSettingsClient(owner?.g_LocalSettingsManager?.Settings?.ManagerIP);
        }

        public void Start()
        {
            Task.Run(StartAsync);
        }

        public void Stop()
        {
            Interlocked.Exchange(ref stopping, 1);
            HubConnection local;

            lock (syncRoot)
            {
                local = connection;
                hubProxy = null;
                connection = null;
            }

            if (local == null)
            {
                return;
            }

            try
            {
                local.Stop();
                local.Dispose();
            }
            catch (Exception ex)
            {
                Logger.WriteErrorLog($"SignalR client stop failed: {ex}", Logger.GetLogFileName());
            }
        }

        public void Reconnect()
        {
            Stop();
            Interlocked.Exchange(ref stopping, 0);
            Interlocked.Exchange(ref reconnecting, 0);
            Start();
        }

        public void Dispose()
        {
            Stop();
            serverSettingsClient?.Dispose();
        }

        private async Task StartAsync()
        {
            bool connected = await EnsureConnectionAsync(forceRefreshSettings: false);
            if (!connected)
            {
                ScheduleReconnect(allowStartWhenDisconnected: true);
            }
        }

        private HubConnection BuildConnection(string url)
        {
            var hub = new HubConnection(url, BuildQueryString());
            hubProxy = hub.CreateHubProxy("MsgHub");
            hubProxy.On<SignalRMessage>("ReceiveMessage", OnReceiveMessage);          
            hub.Closed += OnClosed;
            hub.Error += ex => Logger.WriteErrorLog($"SignalR client error: {ex}", Logger.GetLogFileName());

            return hub;
        }

        private IDictionary<string, string> BuildQueryString()
        {
            var query = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
            var player = owner?.g_PlayerInfoManager?.g_PlayerInfo;
            if (player == null)
            {
                return query;
            }

            if (!string.IsNullOrWhiteSpace(player.PIF_PlayerName))
            {
                query["playerName"] = player.PIF_PlayerName;
            }

            if (!string.IsNullOrWhiteSpace(player.PIF_GUID))
            {
                query["playerGuid"] = player.PIF_GUID;
            }

            return query;
        }

        private void OnClosed()
        {
            if (Interlocked.CompareExchange(ref stopping, 0, 0) == 1)
            {
                return;
            }

            Logger.WriteLog("SignalR client disconnected. Reconnecting...", Logger.GetLogFileName());
            ScheduleReconnect(allowStartWhenDisconnected: false);
        }

        public void SendHeartbeat(HeartbeatPayload payload)
        {
            if (payload == null)
            {
                return;
            }

            HubConnection localConnection;
            IHubProxy localProxy;
            lock (syncRoot)
            {
                localConnection = connection;
                localProxy = hubProxy;
            }

            if (localConnection == null || localProxy == null || localConnection.State != ConnectionState.Connected)
            {
                ScheduleReconnect(allowStartWhenDisconnected: true);
                return;
            }

            Task.Run(async () =>
            {
                try
                {
                    await localProxy.Invoke("ReportHeartbeat", payload);
                }
                catch (Exception ex)
                {
                    Logger.WriteErrorLog($"SignalR heartbeat send failed: {ex.Message}", Logger.GetLogFileName());
                    ScheduleReconnect(allowStartWhenDisconnected: true);
                }
            });
        }

        private void ScheduleReconnect(bool allowStartWhenDisconnected)
        {
            if (Interlocked.Exchange(ref reconnecting, 1) == 1)
            {
                return;
            }

            Task.Run(async () =>
            {
                try
                {
                    while (Interlocked.CompareExchange(ref stopping, 0, 0) == 0)
                    {
                        await Task.Delay(ReconnectDelayMs);
                        HubConnection local;
                        lock (syncRoot)
                        {
                            local = connection;
                        }

                        if (local == null && !allowStartWhenDisconnected)
                        {
                            return;
                        }

                        bool connected = await EnsureConnectionAsync(forceRefreshSettings: true);
                        if (connected)
                        {
                            return;
                        }
                    }
                }
                finally
                {
                    Interlocked.Exchange(ref reconnecting, 0);
                }
            });
        }

        private void OnReceiveMessage(SignalRMessage message)
        {
            if (message == null)
            {
                return;
            }

            if (string.Equals(message.DataType, "CommandQueue", StringComparison.OrdinalIgnoreCase))
            {
                var envelope = ExtractCommandEnvelope(message);
                if (envelope == null)
                {
                    return;
                }

                commandService?.HandleCommandFromSignalR(envelope);
                return;
            }

            if (string.Equals(message.DataType, "StateMessage", StringComparison.OrdinalIgnoreCase))
            {
                return;
            }

            if (string.Equals(message.DataType, "WeeklySchedule", StringComparison.OrdinalIgnoreCase)
                || string.Equals(message.Command, "weekly-schedule-updated", StringComparison.OrdinalIgnoreCase))
            {
                var weekly = ExtractWeeklySchedule(message);
                if (weekly != null)
                {
                    commandService?.HandleWeeklyScheduleFromSignalR(weekly);
                }
                return;
            }

            string command = ExtractCommand(message);
            if (string.IsNullOrWhiteSpace(command))
            {
                return;
            }

            commandService?.HandleCommandFromSignalR(command);
        }

        private SignalRCommandEnvelope ExtractCommandEnvelope(SignalRMessage message)
        {
            if (message == null || message.Data == null)
            {
                return null;
            }

            if (message.Data is SignalRCommandEnvelope envelope)
            {
                return envelope;
            }

            try
            {
                if (message.Data is JObject jObject)
                {
                    return jObject.ToObject<SignalRCommandEnvelope>();
                }

                if (message.Data is string raw)
                {
                    if (string.IsNullOrWhiteSpace(raw))
                    {
                        return null;
                    }

                    return JsonConvert.DeserializeObject<SignalRCommandEnvelope>(raw);
                }

                string json = JsonConvert.SerializeObject(message.Data);
                if (string.IsNullOrWhiteSpace(json))
                {
                    return null;
                }

                return JsonConvert.DeserializeObject<SignalRCommandEnvelope>(json);
            }
            catch (Exception ex)
            {
                Logger.WriteErrorLog($"SignalR command envelope parse failed: {ex.Message}", Logger.GetLogFileName());
                return null;
            }
        }

        private string ExtractCommand(SignalRMessage message)
        {
            string command = message.Command ?? string.Empty;
            string dataCommand = message.Data as string;

            if (string.Equals(command, "Update", StringComparison.OrdinalIgnoreCase))
            {
                if (!string.IsNullOrWhiteSpace(dataCommand))
                {
                    command = dataCommand;
                }
                else
                {
                    command = "updatelist";
                }
            }

            if (string.IsNullOrWhiteSpace(command))
            {
                return null;
            }

            return command.Trim().ToLowerInvariant();
        }

        private SharedWeeklyPlayScheduleInfo ExtractWeeklySchedule(SignalRMessage message)
        {
            if (message == null || message.Data == null)
            {
                return null;
            }

            try
            {
                if (message.Data is SharedWeeklyPlayScheduleInfo weekly)
                {
                    return weekly;
                }

                if (message.Data is JObject jObj)
                {
                    return jObj.ToObject<WeeklyPlayScheduleInfo>();
                }

                if (message.Data is string raw)
                {
                    if (string.IsNullOrWhiteSpace(raw))
                    {
                        return null;
                    }

                    return JsonConvert.DeserializeObject<SharedWeeklyPlayScheduleInfo>(raw);
                }

                string json = JsonConvert.SerializeObject(message.Data);
                if (string.IsNullOrWhiteSpace(json))
                {
                    return null;
                }

                return JsonConvert.DeserializeObject<SharedWeeklyPlayScheduleInfo>(json);
            }
            catch (Exception ex)
            {
                Logger.WriteErrorLog($"SignalR weekly schedule parse failed: {ex.Message}", Logger.GetLogFileName());
                return null;
            }
        }

        private string BuildUrl()
        {
            return BuildUrl(forceRefreshSettings: false);
        }

        private string BuildUrl(bool forceRefreshSettings)
        {
            string rethinkHost = owner?.g_LocalSettingsManager?.Settings?.ManagerIP;
            serverSettingsClient?.UpdateHost(rethinkHost);
            var settings = serverSettingsClient?.GetSettings(forceRefreshSettings);

            string host = settings?.MessageServerIp;
            if (string.IsNullOrWhiteSpace(host))
            {
                Logger.WriteLog("SignalR client skipped: ServerSettings(MessageServerIp) not found in RethinkDB.", Logger.GetLogFileName());
                return null;
            }

            int port = ResolvePort();
            string hubPath = ResolveHubPath();
            if (!hubPath.StartsWith("/"))
            {
                hubPath = "/" + hubPath;
            }

            return $"http://{host}:{port}{hubPath}";
        }

        private async Task<bool> EnsureConnectionAsync(bool forceRefreshSettings)
        {
            string url = BuildUrl(forceRefreshSettings);
            if (string.IsNullOrWhiteSpace(url))
            {
                return false;
            }

            HubConnection local = null;
            HubConnection old = null;
            lock (syncRoot)
            {
                if (connection != null)
                {
                    if (string.Equals(currentUrl, url, StringComparison.OrdinalIgnoreCase))
                    {
                        local = connection;
                    }
                    else
                    {
                        old = connection;
                        hubProxy = null;
                        connection = null;
                        currentUrl = null;
                    }
                }

                if (local == null)
                {
                    local = BuildConnection(url);
                    connection = local;
                    currentUrl = url;
                }
            }

            if (old != null)
            {
                try
                {
                    old.Stop();
                    old.Dispose();
                }
                catch (Exception ex)
                {
                    Logger.WriteErrorLog($"SignalR client stop failed: {ex}", Logger.GetLogFileName());
                }
            }

            if (local.State == ConnectionState.Connected)
            {
                return true;
            }

            try
            {
                await local.Start();
                Logger.WriteLog($"SignalR client connected: {url}", Logger.GetLogFileName());
                return true;
            }
            catch (Exception ex)
            {
                Logger.WriteErrorLog($"SignalR client connect failed: {ex}", Logger.GetLogFileName());
                return false;
            }
        }

        private static int ResolvePort()
        {
            string value = ConfigurationManager.AppSettings["SignalRPort"];
            if (int.TryParse(value, out int port) && port > 0 && port <= 65535)
            {
                return port;
            }

            return DefaultPort;
        }

        private static string ResolveHubPath()
        {
            string value = ConfigurationManager.AppSettings["SignalRHubPath"];
            if (string.IsNullOrWhiteSpace(value))
            {
                return DefaultHubPath;
            }

            return value;
        }
    }

    public class SignalRMessage
    {
        public string From { get; set; } = "Server";

        public string To { get; set; } = "All";

        public string Command { get; set; } = "Update";

        public string DataType { get; set; } = "String";

        public object Data { get; set; } = null;
    }

    public class SignalRCommandEnvelope
    {
        public string CommandId { get; set; }
        public string Command { get; set; }
        public string PlayerId { get; set; }
        public string PayloadJson { get; set; }
        public string CreatedAt { get; set; }
        public bool IsUrgent { get; set; }
    }

    public class StateMessage
    {
        public string Who { get; set; } = "Unknown";

        public string State { get; set; } = "Disconnected";

        public string Description { get; set; } = "";
    }

    public class ProgressData
    {
        public string FromGUID { get; set; }

        public int Porgress { get; set; }
    }
}

File: Player/Android/app/src/main/java/kr/co/turtlelab/andowsignage/services/SignalRClientService.java
package kr.co.turtlelab.andowsignage.services;

import android.text.TextUtils;
import android.util.Log;

import com.google.gson.Gson;

import java.net.URLEncoder;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import kr.co.turtlelab.andowsignage.AndoWSignageApp;
import kr.co.turtlelab.andowsignage.dataproviders.LocalSettingsProvider;
import kr.co.turtlelab.andowsignage.data.rethink.RethinkDbClient;
import kr.co.turtlelab.andowsignage.data.rethink.RethinkModels;
import microsoft.aspnet.signalr.client.ConnectionState;
import microsoft.aspnet.signalr.client.NullLogger;
import microsoft.aspnet.signalr.client.Platform;
import microsoft.aspnet.signalr.client.SignalRFuture;
import microsoft.aspnet.signalr.client.hubs.HubConnection;
import microsoft.aspnet.signalr.client.hubs.HubProxy;
import microsoft.aspnet.signalr.client.hubs.SubscriptionHandler1;
import microsoft.aspnet.signalr.client.http.android.AndroidPlatformComponent;
import microsoft.aspnet.signalr.client.transport.LongPollingTransport;

public class SignalRClientService {
    private static final String TAG = "SignalRClientService";

    public interface Listener {
        void onCommand(String command);
        void onCommandEnvelope(SignalRCommandEnvelope envelope);
        void onWeeklySchedule(RethinkModels.WeeklyScheduleRecord weekly);
    }

    private static final int DEFAULT_PORT = 5000;
    private static final String DEFAULT_HUB_PATH = "/Data";
    private static final String HUB_NAME = "MsgHub";
    private static final long RECONNECT_DELAY_MS = 5000L;
    private static final long CONNECT_TIMEOUT_MS = 15000L;
    private static final long HEARTBEAT_TIMEOUT_MS = 5000L;
    private static final int MAX_SIGNALR_ACTION_QUEUE = 10;
    private static final NullLogger SIGNALR_NULL_LOGGER = new NullLogger();

    private Listener listener;
    private final Gson gson;
    private final ThreadPoolExecutor actionExecutor = new ThreadPoolExecutor(
            1,
            1,
            0L,
            TimeUnit.MILLISECONDS,
            new LinkedBlockingDeque<>(MAX_SIGNALR_ACTION_QUEUE),
            runnable -> {
                Thread thread = new Thread(runnable, "SignalRActionQueue");
                thread.setDaemon(true);
                return thread;
            });
    private final ScheduledExecutorService reconnectScheduler = Executors.newSingleThreadScheduledExecutor(runnable -> {
        Thread thread = new Thread(runnable, "SignalRReconnectScheduler");
        thread.setDaemon(true);
        return thread;
    });
    private final Object syncRoot = new Object();
    private final AtomicBoolean stopping = new AtomicBoolean(false);
    private final AtomicBoolean reconnectScheduled = new AtomicBoolean(false);

    private HubConnection connection;
    private HubProxy hubProxy;
    private ScheduledFuture<?> reconnectFuture;

    private static SignalRClientService shared;

    public SignalRClientService(Listener listener, Gson gson) {
        this.listener = listener;
        this.gson = gson == null ? new Gson() : gson;
        Platform.loadPlatformComponent(new AndroidPlatformComponent());
        actionExecutor.setRejectedExecutionHandler((runnable, executor) -> {
            if (executor.isShutdown()) {
                throw new RejectedExecutionException("SignalR action queue is shut down");
            }
            Runnable dropped = executor.getQueue().poll();
            Log.w(TAG, "SignalR action queue full. dropped oldest action=" + describeAction(dropped));
            executor.execute(runnable);
        });
    }

    public static synchronized SignalRClientService getShared(Gson gson) {
        if (shared == null) {
            shared = new SignalRClientService(null, gson);
        }
        return shared;
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    public void start() {
        stopping.set(false);
        enqueueAction("start", this::ensureConnectedInternal);
    }

    public void stop() {
        stopping.set(true);
        cancelReconnectSchedule();
        enqueueAction("stop", this::stopInternal);
    }

    public void reconnect() {
        stopping.set(false);
        cancelReconnectSchedule();
        enqueueAction("reconnect", () -> {
            stopInternal();
            ensureConnectedInternal();
        });
    }

    public void sendHeartbeat(HeartbeatPayload payload) {
        if (payload == null) {
            return;
        }
        enqueueAction("heartbeat", () -> sendHeartbeatInternal(payload));
    }

    private void ensureConnectedInternal() {
        if (stopping.get()) {
            return;
        }
        String url = buildUrl();
        if (TextUtils.isEmpty(url)) {
            Log.w(TAG, "ensureConnectedInternal: empty SignalR url");
            return;
        }

        HubConnection existingConnection;
        synchronized (syncRoot) {
            existingConnection = connection;
            if (shouldKeepCurrentConnection(existingConnection)) {
                return;
            }
        }

        ConnectionBundle bundle = buildConnection(url);
        HubConnection previous;
        synchronized (syncRoot) {
            previous = connection;
            connection = bundle.connection;
            hubProxy = bundle.proxy;
        }
        safeStopConnection(previous);

        try {
            startConnection(bundle.connection);
            reconnectScheduled.set(false);
        } catch (Exception ex) {
            if (isOperationCancelled(ex)) {
                Log.w(TAG, "startInternal: startConnection cancelled. url=" + url, ex);
            } else {
                Log.e(TAG, "startInternal: startConnection failed, scheduling reconnect. url=" + url, ex);
            }
            synchronized (syncRoot) {
                if (connection == bundle.connection) {
                    connection = null;
                    hubProxy = null;
                }
            }
            safeStopConnection(bundle.connection);
            if (!stopping.get()) {
                scheduleReconnect();
            }
        }
    }

    private ConnectionBundle buildConnection(String url) {
        HubConnection hubConnection = new HubConnection(url, buildQueryString(), true, SIGNALR_NULL_LOGGER);
        HubProxy proxy = hubConnection.createHubProxy(HUB_NAME);
        proxy.on("ReceiveMessage", new SubscriptionHandler1<SignalRMessage>() {
            @Override
            public void run(SignalRMessage message) {
                handleMessage(message);
            }
        }, SignalRMessage.class);
        hubConnection.closed(() -> {
            if (!stopping.get()) {
                synchronized (syncRoot) {
                    if (connection == hubConnection) {
                        connection = null;
                        hubProxy = null;
                    }
                }
                scheduleReconnect();
            }
        });
        return new ConnectionBundle(hubConnection, proxy);
    }

    private void scheduleReconnect() {
        if (stopping.get()) {
            return;
        }
        if (reconnectScheduled.getAndSet(true)) {
            return;
        }
        synchronized (syncRoot) {
            if (reconnectFuture != null) {
                reconnectFuture.cancel(false);
            }
            reconnectFuture = reconnectScheduler.schedule(() -> {
                reconnectScheduled.set(false);
                if (!stopping.get()) {
                    enqueueAction("scheduled-reconnect", this::ensureConnectedInternal);
                }
            }, RECONNECT_DELAY_MS, TimeUnit.MILLISECONDS);
        }
    }

    private void safeStopConnection(HubConnection hubConnection) {
        if (hubConnection == null) {
            return;
        }
        try {
            hubConnection.stop();
        } catch (Exception ex) {
            Log.w(TAG, "safeStopConnection: failed to stop hub connection", ex);
        }
    }

    private void stopInternal() {
        HubConnection local;
        synchronized (syncRoot) {
            local = connection;
            connection = null;
            hubProxy = null;
        }
        safeStopConnection(local);
    }

    private static final class ConnectionBundle {
        final HubConnection connection;
        final HubProxy proxy;

        ConnectionBundle(HubConnection connection, HubProxy proxy) {
            this.connection = connection;
            this.proxy = proxy;
        }
    }

    private void startConnection(HubConnection local) throws Exception {
        // 이전 작업의 interrupt 플래그가 남아 있으면 future.get()가 즉시 취소될 수 있어 초기화한다.
        if (Thread.interrupted()) {
            Log.w(TAG, "startConnection: clearing stale interrupted flag before connect.");
        }
        SignalRFuture<Void> future = local.start(new LongPollingTransport(SIGNALR_NULL_LOGGER));
        future.get(CONNECT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
    }

    private boolean shouldKeepCurrentConnection(HubConnection local) {
        if (local == null) {
            return false;
        }
        ConnectionState state = local.getState();
        return state == ConnectionState.Connected || state == ConnectionState.Connecting;
    }

    private void sendHeartbeatInternal(HeartbeatPayload payload) {
        if (stopping.get()) {
            return;
        }
        HubConnection localConn;
        HubProxy localProxy;
        synchronized (syncRoot) {
            localConn = connection;
            localProxy = hubProxy;
        }
        if (localConn == null || localProxy == null) {
            scheduleReconnect();
            return;
        }
        if (localConn.getState() != ConnectionState.Connected) {
            scheduleReconnect();
            return;
        }
        try {
            SignalRFuture<Void> future = localProxy.invoke("ReportHeartbeat", payload);
            future.get(HEARTBEAT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
        } catch (Exception ex) {
            Log.e(TAG, "sendHeartbeat: ReportHeartbeat invoke failed", ex);
            synchronized (syncRoot) {
                if (connection == localConn) {
                    connection = null;
                    hubProxy = null;
                }
            }
            safeStopConnection(localConn);
            scheduleReconnect();
        }
    }

    private void cancelReconnectSchedule() {
        reconnectScheduled.set(false);
        synchronized (syncRoot) {
            if (reconnectFuture != null) {
                reconnectFuture.cancel(false);
                reconnectFuture = null;
            }
        }
    }

    private void enqueueAction(String label, Runnable action) {
        try {
            actionExecutor.execute(new NamedAction(label, action));
        } catch (RejectedExecutionException ex) {
            Log.e(TAG, "enqueueAction failed. label=" + label, ex);
        }
    }

    private static String describeAction(Runnable runnable) {
        if (runnable instanceof NamedAction) {
            return ((NamedAction) runnable).label;
        }
        return runnable == null ? "none" : runnable.getClass().getSimpleName();
    }

    private static final class NamedAction implements Runnable {
        private final String label;
        private final Runnable delegate;

        private NamedAction(String label, Runnable delegate) {
            this.label = label;
            this.delegate = delegate;
        }

        @Override
        public void run() {
            delegate.run();
        }
    }

    private boolean isOperationCancelled(Throwable throwable) {
        Throwable current = throwable;
        while (current != null) {
            if (current instanceof InterruptedException) {
                return true;
            }
            String message = current.getMessage();
            if (!TextUtils.isEmpty(message) && message.toLowerCase(Locale.US).contains("operation was cancelled")) {
                return true;
            }
            current = current.getCause();
        }
        return false;
    }

    private void handleMessage(SignalRMessage message) {
        if (message == null) {
            return;
        }

        if ("CommandQueue".equalsIgnoreCase(message.DataType)) {
            SignalRCommandEnvelope envelope = extractCommandEnvelope(message);
            if (envelope != null && listener != null) {
                listener.onCommandEnvelope(envelope);
            }
            return;
        }

        if ("StateMessage".equalsIgnoreCase(message.DataType)) {
            return;
        }

        if ("WeeklySchedule".equalsIgnoreCase(message.DataType)
                || "weekly-schedule-updated".equalsIgnoreCase(message.Command)) {
            RethinkModels.WeeklyScheduleRecord weekly = extractWeeklySchedule(message);
            if (weekly != null && listener != null) {
                listener.onWeeklySchedule(weekly);
            }
            return;
        }

        String command = extractCommand(message);
        if (!TextUtils.isEmpty(command) && listener != null) {
            listener.onCommand(command);
        }
    }

    private SignalRCommandEnvelope extractCommandEnvelope(SignalRMessage message) {
        if (message == null || message.Data == null) {
            return null;
        }
        try {
            if (message.Data instanceof SignalRCommandEnvelope) {
                return (SignalRCommandEnvelope) message.Data;
            }
            if (message.Data instanceof String) {
                String raw = (String) message.Data;
                if (TextUtils.isEmpty(raw)) {
                    return null;
                }
                return gson.fromJson(raw, SignalRCommandEnvelope.class);
            }
            String json = gson.toJson(message.Data);
            if (TextUtils.isEmpty(json)) {
                return null;
            }
            return gson.fromJson(json, SignalRCommandEnvelope.class);
        } catch (Exception ex) {
            Log.w(TAG, "extractCommandEnvelope: failed to parse message data", ex);
            return null;
        }
    }

    private String extractCommand(SignalRMessage message) {
        if (message == null) {
            return null;
        }
        String command = message.Command == null ? "" : message.Command;
        String dataCommand = null;
        if (message.Data instanceof String) {
            dataCommand = (String) message.Data;
        }
        if ("Update".equalsIgnoreCase(command)) {
            if (!TextUtils.isEmpty(dataCommand)) {
                command = dataCommand;
            } else {
                command = "updatelist";
            }
        }
        if (TextUtils.isEmpty(command)) {
            return null;
        }
        return command.trim().toLowerCase(Locale.US);
    }

    private RethinkModels.WeeklyScheduleRecord extractWeeklySchedule(SignalRMessage message) {
        if (message == null || message.Data == null) {
            return null;
        }
        try {
            if (message.Data instanceof RethinkModels.WeeklyScheduleRecord) {
                return (RethinkModels.WeeklyScheduleRecord) message.Data;
            }
            if (message.Data instanceof String) {
                String raw = (String) message.Data;
                if (TextUtils.isEmpty(raw)) {
                    return null;
                }
                return gson.fromJson(raw, RethinkModels.WeeklyScheduleRecord.class);
            }
            String json = gson.toJson(message.Data);
            if (TextUtils.isEmpty(json)) {
                return null;
            }
            return gson.fromJson(json, RethinkModels.WeeklyScheduleRecord.class);
        } catch (Exception ex) {
            Log.w(TAG, "extractWeeklySchedule: failed to parse weekly payload", ex);
            return null;
        }
    }

    private String buildUrl() {
        String host = resolveHost();
        if (TextUtils.isEmpty(host)) {
            return null;
        }
        int port = resolvePort();
        String hubPath = resolveHubPath();
        if (!hubPath.startsWith("/")) {
            hubPath = "/" + hubPath;
        }
        return "http://" + host + ":" + port + hubPath;
    }

    private String buildQueryString() {
        try {
            String playerName = RethinkDbClient.getInstance().getStoredPlayerName();
            if (TextUtils.isEmpty(playerName)) {
                playerName = AndoWSignageApp.PLAYER_ID;
            }
            String playerGuid = RethinkDbClient.getInstance().getStoredPlayerGuid();
            StringBuilder sb = new StringBuilder();
            if (!TextUtils.isEmpty(playerName)) {
                sb.append("playerName=").append(URLEncoder.encode(playerName, "UTF-8"));
            }
            if (!TextUtils.isEmpty(playerGuid)) {
                if (sb.length() > 0) sb.append("&");
                sb.append("playerGuid=").append(URLEncoder.encode(playerGuid, "UTF-8"));
            }
            return sb.toString();
        } catch (Exception ex) {
            Log.e(TAG, "buildQueryString: failed to build query string", ex);
            return "";
        }
    }

    private String resolveHost() {
        String messageServerIp = LocalSettingsProvider.getMessageServerIp();
        if (!TextUtils.isEmpty(messageServerIp)) {
            return messageServerIp;
        }
        if (AndoWSignageApp.IS_MANUAL && !TextUtils.isEmpty(AndoWSignageApp.MANUAL_IP)) {
            return AndoWSignageApp.MANUAL_IP;
        }
        String managerIp = LocalSettingsProvider.getManagerIp();
        if (!TextUtils.isEmpty(managerIp)) {
            return managerIp;
        }
        return AndoWSignageApp.MANAGER_IP;
    }

    private int resolvePort() {
        int configured = LocalSettingsProvider.getSignalrPort();
        if (configured > 0 && configured <= 65535) {
            return configured;
        }
        return DEFAULT_PORT;
    }

    private String resolveHubPath() {
        String configured = LocalSettingsProvider.getSignalrHubPath();
        if (TextUtils.isEmpty(configured)) {
            return DEFAULT_HUB_PATH;
        }
        return configured;
    }

    public static class SignalRMessage {
        public String From = "Server";
        public String To = "All";
        public String Command = "Update";
        public String DataType = "String";
        public Object Data = null;
    }

    public static class SignalRCommandEnvelope {
        public String CommandId;
        public String Command;
        public String PlayerId;
        public String PayloadJson;
        public String CreatedAt;
        public boolean IsUrgent;
    }

    public static class HeartbeatPayload {
        public String ClientId = "";
        public String PlayerName = "";
        public String Status = "";
        public int Process;
        public String Version = "";
        public String CurrentPage = "";
        public boolean HdmiState;

        public static HeartbeatPayload create(String clientId,
                                              String status,
                                              int process,
                                              String version,
                                              String currentPage,
                                              boolean hdmiState) {
            HeartbeatPayload payload = new HeartbeatPayload();
            payload.ClientId = clientId == null ? "" : clientId;
            payload.PlayerName = AndoWSignageApp.PLAYER_ID == null ? "" : AndoWSignageApp.PLAYER_ID;
            payload.Status = status == null ? "" : status;
            payload.Process = process;
            payload.Version = version == null ? "" : version;
            payload.CurrentPage = currentPage == null ? "" : currentPage;
            payload.HdmiState = hdmiState;
            return payload;
        }
    }
}
